Implement API for Shulkers, ShulkerBullets and ShulkerBoxes

RedNesto opened the pull request:
[SpongeAPI](https://github.com/SpongePowered/SpongeAPI/pull/1903) | **SpongeCommon**

This PR implements everything added to its SpongeAPI counterpart.

I hope I did all the DataAPI stuff right.

Reviews are welcome.

_Only tested with SpongeVanilla for now._

[Test Plugin](https://gist.github.com/RedNesto/6fdd0d1e8ce66a52374432e1544c43ba)

RedNesto committed:
Implement API according additions made to Shulkers, ShulkerBullets and ShulkerBoxes

XakepSDK commented on src/main/java/org/spongepowered/common/data/processor/data/entity/EntityTargetedEntityDataProcessor.java:
```diff
@@ -0,0 +1,88 @@
+
+    @Override
+    protected boolean set(Entity dataHolder, Optional<org.spongepowered.api.entity.Entity> value) {
+        if (!(dataHolder instanceof IEntityTargetingEntity))
```
Braces

XakepSDK commented on src/main/java/org/spongepowered/common/data/processor/data/entity/EntityTargetedEntityDataProcessor.java:
```diff
@@ -0,0 +1,88 @@
+
+    @Override
+    protected Optional<Optional<org.spongepowered.api.entity.Entity>> getVal(Entity dataHolder) {
+        if (!(dataHolder instanceof IEntityTargetingEntity))
```
Braces

XakepSDK commented on src/main/java/org/spongepowered/common/data/processor/data/entity/ShulkerDyeColorDataProcessor.java:
```diff
@@ -0,0 +1,87 @@
+
+    @Override
+    protected boolean set(EntityShulker container, DyeColor value) {
+        if (!(container instanceof IMixinShulker))
```
Braces

XakepSDK commented on src/main/java/org/spongepowered/common/data/processor/data/entity/ShulkerDyeColorDataProcessor.java:
```diff
@@ -0,0 +1,87 @@
+
+    @Override
+    protected Optional<DyeColor> getVal(EntityShulker container) {
+        if (!(container instanceof IMixinShulker))
```
Braces

RedNesto committed:
Add missing braces and remove unused things

ImMorpheus commented on src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntityShulkerBullet.java:
```diff
@@ -24,11 +24,115 @@
+    @Override
+    public void supplyVanillaManipulators(List<DataManipulator<?, ?>> manipulators) {
+        super.supplyVanillaManipulators(manipulators);
+        manipulators.add(new SpongeTargetedEntityData());
```
This is wrong. It's always returning an empty target.

RedNesto committed:
Do DataProcessors sanity checks in #supports
Give a full name to ShulkerDirectionalProcessor
Use ShulkerBullet#getTargetData in supplyVanillaManipulators

RedNesto committed:
Override the right #supports method

ImMorpheus commented on src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntityShulkerBullet.java:
```diff
@@ -24,11 +24,115 @@
+
+    @Override
+    public ProjectileSource getShooter() {
+        if (projectileSource != null) {
```
this

RedNesto committed:
Add a missing this qualifier

RedNesto committed:
Make Keys#TARGETED_ENTITY a simple Value<> and replace ShulkerBullet#getDirection by a DirectionalData

ImMorpheus reviewed (changes requested):
As in the signsource PR, add the test plugin source as a proper test plugin.

RedNesto committed:
Add test plugin for shulkers

ImMorpheus committed:
Entity -> EntitySnapshot

ImMorpheus committed:
Use the correct TypeToken

ImMorpheus committed:
Update testplugin

ImMorpheus committed:
License

ImMorpheus committed:
Actually implement directional data
